Minimax

td-extreme opened the pull request.

td-extreme committed:
Added PlayerAiMinimax class. Created first tests.

td-extreme committed:
Created PlayerAiMiniax class.

td-extreme committed:
Cleaned up white space.

dadarek commented on lib/player_ai_minimax.rb:
```diff
@@ -0,0 +1,57 @@
+  private
+  def opponent_played_corner_for_first_move(board)
+    corners = [0,2,6,8]
+    if (board.available_moves.size == 8)
```
look at the Enumerable module for the `any?` method.

Could probably do something like: `corners.any?{ |corner| board.spaces[corner] != ' ' }`

dadarek commented on lib/player_ai_minimax.rb:
```diff
@@ -0,0 +1,57 @@
+    return 0
+  end
+
+  def minimax(board, player, opponent)
```
This function is a little chunky. Two big things going on here that could probably be broken up into helper fns.

dadarek commented on lib/player_ai_minimax.rb:
```diff
@@ -0,0 +1,57 @@
+    end
+    if player == @current_player
+      max_score_index = scores.each_with_index.max[1]
+      @choice = moves[max_score_index]
```
It _looks_ to me like this instance variable is being set only so that it can be returned on line 16. This is not ideal since it does not follow very easily (kind of have to bounce back and forth) and also creates this side-effect dependency that isn't immediately obvious.

It feels like a command/query violation: i.e. the minimax fn is doing some calculation, _but_ also it's setting this instance variable because it knows that some function up in the call stack will expect this variable to be set.

dadarek commented on spec/player_ai_minimax_spec.rb:
```diff
@@ -0,0 +1,108 @@
+require 'game_rules'
+
+describe PlayerAiMinimax do
+  let (:io) { double }
```
Is this used?

dadarek commented on spec/player_ai_minimax_spec.rb:
```diff
@@ -0,0 +1,108 @@
+  
+  describe "Playing the last move " do
+    it "plays 0 when that is the only space open " do
+      [1,2,3,4,5,6,7,8].each do |i|
```
FYI: In ruby you could do something like (not 100% sure on the syntax but): `(1..8).each` or `(1...8)` for inclusive. Lookup ruby ranges.

dadarek commented on spec/player_ai_minimax_spec.rb:
```diff
@@ -0,0 +1,108 @@
+  end
+
+  describe "Playing second to last move " do
+    it "when only 7 and 8 are left it plays 7 when that will block opponent and 8 will not " do
```
just a nitpick - wording is confusing here. Not a big deal, but coming up with good descriptions is important, especially when that day comes when you're trying to understand a test that somebody wrote that is super-confusing to follow and the description offers absolutely _no_ help.

dadarek commented on spec/player_ai_minimax_spec.rb:
```diff
@@ -0,0 +1,108 @@
+      [1,2,3,4].each do |i|
+        spaces[i] = :X
+      end
+      [0,5,6].each do |i|
```
Not sure if you know this but gonna mention it anyway: in ruby you can do single-line blocks for simple things. i.e.
`[0,5,6].each { |i| spaces[i] = :O }`

dadarek commented on spec/player_ai_minimax_spec.rb:
```diff
@@ -0,0 +1,108 @@
+    end
+
+    it "when only 1 and 2 are left it plays 2 and neither move wins / blocks " do
+      [0,4,5,6].each do |i|
```
This idea of "let's put x on these spaces and O on these spaces" seems to be a recurring thing. Might be worth ripping out a helper fn to help the tests read a little better.

dadarek commented on spec/player_ai_minimax_spec.rb:
```diff
@@ -0,0 +1,108 @@
+      expect(test_ai.play_move(test_state)).to eq(7)
+    end
+
+    it "when only 1 and 2 are left it plays 2 and neither move wins / blocks " do
```
Wait, why isn't playing a 2?

dadarek commented on spec/player_ai_minimax_spec.rb:
```diff
@@ -0,0 +1,108 @@
+    end
+  end
+
+  describe "When it goes second " do
```
How about - if it's going second and the other player did _not_ play a corner?

dadarek commented on spec/player_ai_minimax_spec.rb:
```diff
@@ -0,0 +1,108 @@
+  end
+
+  describe "Creating a Fork " do
+    it "plays 2 to create a fork " do
```
With minimax being as complicated as it is, throwing in another fork test here probably wouldn't hurt. Choose a setup where the fork is created by playing something other than 2.

dadarek commented:
In addition to the other comments I already left, it's probably valuable to have:
- a few tests for the best third move. (not sure what that would be but I feel like there are "better" third moves and "worse" third moves.
- a few tests for "in the middle" game play where the AI takes a winning move on the fourth or fifth turn and also prevents a winning move (i.e. blocks) on the fourth or fifth turn.

They don't have to be exhaustive, but right now we only have tests for the last move, for the first move (which is hardcoded), limited second-move tests (which are also hardcoded) and one test for a fork, which are probably a bit too light for something this complicated.

dadarek commented on spec/player_ai_minimax_spec.rb:
```diff
@@ -0,0 +1,108 @@
+  let (:test_ai2) { PlayerAiMinimax.new(:O, rules) }
+  let (:test_state) { GameState.new(board , :X, :O) }
+  let(:spaces) { test_state.game_board.spaces }
+  let(:corners) { [0, 2, 6, 8] }
```
unused?

td-extreme committed:
Cleaned up tests & added new Tests.

Made a helper function for playing mulitple moves on the board, as this
was used in the setup of several tests.
Added tests to see if ai goes second and player 1 played the center or a
side move.

td-extreme committed:
Added hard logic for first turn if it is player 2.

Instead of using minimax on its first turn if it is second player it
plays the center if a cornere was played and a corner if the center or a
side move was played by the first player.

mikeknep commented on lib/player_ai_minimax.rb:
```diff
@@ -0,0 +1,60 @@
+  def corner_move(board)
+    available_corners = @corners & board.available_moves
+    i = Random.rand(available_corners.size)
+    available_corners[i]
```
Nitpick: you could consolidate these two lines by using [#sample](http://ruby-doc.org/core-2.2.0/Array.html#method-i-sample)

mikeknep commented on lib/player_ai_minimax.rb:
```diff
@@ -0,0 +1,60 @@
+
+  def initialize(mark, rules)
+    @mark = mark
+    @rules = rules
```
You can define `attr_reader`s as `private`. Here's one nice feature of using readers over instance variables throughout your class:

```
class Foo
  def initialize(bar, baz)
    @bar = bar
    @baz = baz
  end

  def get_bar
    @bat   # misspelled @bar instance variable
  end

  def get_baz
    bax   # misspelled baz reader method
  end

  private
  attr_reader :baz
end


f = Foo.new(1,2)
f.get_bar  # => nil.
           # It's unclear whether @bar didn't get set or if the instance variable was just misspelled somewhere

f.get_baz  # => NameError: undefined local variable or method `bax` for #<Foo>.
           # Ah! It's clear I misspelled `baz` somewhere
```

I also happen to think it looks cleaner than having instance variables everywhere...

mikeknep commented on lib/player_ai_minimax.rb:
```diff
@@ -0,0 +1,60 @@
+  def initialize(mark, rules)
+    @mark = mark
+    @rules = rules
+    @corners = [0,2,6,8]
```
Setting aside the fact that this is hardcoded, this could be moved to either a constant or its own method. For example

```
class PlayerAIMinimax
  CORNERS = [0,2,6,8]           # as a constant

  def initialize(mark, rules)
    @mark = mark
    @rules = rules
  end

  # ...

  private

  def corners                   # as a method
    @corners ||= [0,2,6,8]
  end
end
```

Main advantage IMO is it keeps the `initialize` method focused on what needs to be passed in to create the instance.

mikeknep commented on lib/player_ai_minimax.rb:
```diff
@@ -0,0 +1,60 @@
+    board = game_state.game_board
+    return 0 if board.empty?
+    return 4 if opponent_played_corner_for_first_move(board)
+    return corner_move(board) if board.available_moves.size == 8
```
Is this just a performance boost for the first move?

mikeknep replied:
Lines 12-14 are all about optimizing the AI player's first move. Maybe move to one private method like `handle_first_move` to make this top-level / public method a little clearer to follow? It seems kinda noisy currently.

mikeknep commented on bin/ttt:
```diff
@@ -7,13 +7,15 @@ require 'game_io'
 
 p1 = PlayerHuman.new('X', io)
-p2 = PlayerAiBasic.new('O', GameRules.new)
+#p2 = PlayerAiBasic.new('O', GameRules.new)
```
Remember that you can always look back in the git history for a previous implementation—no need to commit this comment.

td-extreme committed:
Added test, updated methods.

Added some additonal tests. Started breaking minimax into more helper
methods. Hopefully should have it cleaned up more tomorrow and be done.

td-extreme committed:
Broke added helper function to minimax method.

td-extreme committed:
Removed line of dead code.

td-extreme committed:
Removed blank lines.

mikeknep commented:
I like the cleanup so far around `first_turn` readability. However, currently your AI will check all the first move options in `first_turn` on _every_ turn, which is unnecessary—it only needs to dive into that logic if it is the first turn. If the first turn logic was _also_ computationally heavy, you'd incur that cost on every turn needlessly.

I think it might be better to have two methods that more cleanly separate command and query: `first_move?` could be a boolean check on whether or not it is currently the first move for the player—should should only need to look at the spots played on the board to determine this. Then you'd have a `select_first_move`, which would be like your current `first_turn`.

Ultimately this can help you get rid of the `||=` (which seem a little awkward here) by refactoring towards something like:

``` ruby
first_move? ? select_first_move : best_move
```

You may not have seen the ternary operator before. The pattern is 

```
predicate_method ? call_when_true : call_when_false
```

In other words, the above code is exactly the same as:

``` ruby
if first_move?
  select_first_move
else
  best_move
end
```

mikeknep commented on lib/player_ai_minimax.rb:
```diff
@@ -0,0 +1,82 @@
+    move ||= corner_move(board) if board.empty?
+    move ||= 4 if opponent_played_corner_for_first_move(board)
+    move ||= corner_move(board) if board.available_moves.size == 8
+    move
```
As I somewhat alluded to in my other comment just now, the use of `||=` here is a little strange. Typically you see it used to cache some computationally heavy or long-running process. For example, if you had a method that returned a parsed JSON output from some API call and you only want to make that call once, not every time you need the value, you could do:

``` ruby
def api_output
  @api_output ||= JSON.parse(API.get(whatever))
end
```

Furthermore, it's a little strange to explicitly set a value to `nil` with the intent to overwrite it. Sometimes it's necessary in Ruby, but in this instance I think a simple `if/elsif/elsif` (falling back on the implicit `nil` if no conditions are met) is more straightforward.

All that being said, I do like how you're thinking of this moving through a pipeline of options/operations—this kind of thinking will be especially valuable in functional languages like Clojure.

mikeknep commented on lib/player_ai_minimax.rb:
```diff
@@ -0,0 +1,82 @@
+      return scores[max_score_index]
+    else
+      min_score_index = scores.each_with_index.min[1]
+      return scores[min_score_index]
```
This `return` and the one in the `if` clause above are unnecessary—an `if/else` block will return the last line of whichever clause was entered/executed.

td-extreme replied:
I updated this method to look like this. 

```
score_index = (player == @current_player) ? \
      scores.each_with_index.max[1] : scores.each_with_index.min[1]
  scores[score_index]
```

mikeknep commented on lib/player_ai_minimax.rb:
```diff
@@ -0,0 +1,82 @@
+  def score(board, depth)
+    return 10 - depth if rules.winner(board) == @current_player
+    return depth - 10  if rules.winner(board) == @non_current_player
+    return 0 - depth
```
`if/elsif/else` is more idiomatic than three `return` statements with predicate checks at the end. In general explicit `return` statements are quite rare in Ruby, and generally to me signal some kind of "early exit," for example:

``` ruby
def foo
  return [] if we_dont_need_to_do_this?
  do_a_whole
  bunch_of_stuff
  that_we_dont
  want_to_do
  if_we_dont_have_to
end
```

td-extreme replied:
Updated this method to :

```
    case rules.winner(board)
    when @current_player
      score = 10 - depth
    when @non_current_player
      score = depth - 10
    else
      score = 0 - depth
    end
    score
```

td-extreme committed:
Added second_turn method.

The minimax algorithm would not always setup for a fork so I hard coded
the logic to make that happen on its second turn.

td-extreme committed:
Cleaned up code.

td-extreme committed:
Cleaned up first_turn method.

dadarek commented on spec/player_ai_minimax_spec.rb:
```diff
@@ -0,0 +1,321 @@
+    end
+
+    it "plays 8 when that is the only space open " do
+      (1...8).each { |i| spaces[i] = i }
```
Can you use the `set_spaces` fn here?

td-extreme replied:
yeah, I could do that. I'll make that change real quick.

td-extreme replied:
made that change for those 3 tests.

dadarek commented:
LGTM 👍

td-extreme committed:
Chagned setup in some test to use helper method.

mikeknep commented on lib/player_ai_minimax.rb:
```diff
@@ -0,0 +1,116 @@
+    elsif second_move?(board)
+      move = second_turn(board)
+    end
+    move ||= best_move(board, game_state.player, game_state.opponent)
```
No need to set a local `move` variable here—this can just be:

``` ruby
if first_move?(board)
  first_turn(board)
elsif second_move?(board)
  second_turn(board)
else
  best_move(board, game_state.player, game_state.opponent)
end
```

mikeknep commented:
I'm not a big fan of the hardcoded second move. It's fairly difficult to read, and while I realize it'll speed up specs and gameplay, I feel like the minimax shouldn't be that much slower when 2 moves have already been played on a 3x3 board. I don't think the performance gain is worth the noise.

I tried commenting out lines 17 and 18, hoping that I'd just fall back down to let minimax choose the right move, and some specs broke. However, while playing the game with those lines out, it appears the minimax computer still chooses successful moves. So I'm guessing you just have some additional tests around the second turn rules that aren't relevant without those rules in place.

I think I'd like to see that just taken out. They'd be fine if we had been making a rules- / logic-based AI, but for a specifically MiniMax AI it doesn't seem appropriate.

td-extreme committed:
Updated play_move method and 1 test.

td-extreme commented:
Mike, 
The second move method is in place because the minimax algorithm wouldn't always attempt to cause a fork. I'm not sure why this happened. I think it is because there would be another way to possibly win without creating a fork. So, I hard coded the second turn to ensure the AI would attempt to cause a fork every time it could. 
I think it is like you said, I was trying to use a more strict set of rules for a logic based AI.

I had two thoughts moving forward on this. 
A) Remove the hardcoded second move and accept that it won't be aggressive in trying to create a fork. This would require removing the tests that are more geared toward a rules / logic-based AI.
B) Leave the hard coded method in place for now, and add todo comment to come back later and figure out if I can make the minimax function actively push for a fork. After fixing the minimax to do that, the hard coded second move could be removed. 

I think option A would probably be easiest for now. Let me know if you agree and I will make those changes.

mikeknep commented:
You're correct. The MiniMax algorithm doesn't always play the most aggressive move to ensure a win in as few moves as possible. It is more conservative. I don't think there's anything wrong with it not pursuing the fork. (In fact, when we review TTTs we are often so used to minimax NOT pursuing a fork that we get surprised when we see one, and it's usually a clue that there's more rule based stuff going on.)

Go ahead with option A, and when that's finished I think this'll be good to go.

td-extreme committed:
Removed tests and logic that was rules based.

mikeknep commented:
Some minor, non-blocker style nitpicks:
- Local variable `move` unnecessary in `#play_move`
- Local variable `score` unnecessary in `#score` (`case` statements return like `if/else` statements)
- `#opponent_played_corner_for_first_move` should be named with a `?` at the end to indicate boolean return type
- parens around `opponent_played_corner_for_first_move` in `#first_turn` are unnecessary
- ternary operator in `#best_score` is kinda hard to read since it's so long that it requires multiple lines. I would probably either use `if/else`, or use a ternary to switch on `:min/:max` and `send` those to `scores.each_with_index`, ex:

``` ruby
def best_score(scores, player)
  min_or_max = (player == @current_player) ? :max : :min
  score_index = scores.each_with_index.send(min_or_max)[1]
  scores[score_index]
end
```

dadarek commented:
`#send` makes developers cry

td-extreme commented:
Mike, I made those nitpick changes and committed it to the master branch, since I had already merged the pull request.

td-extreme commented:
Darek,   what do you mean by your last comment??

> #send makes developers cry

dadarek commented:
Sorry @td-extreme just saw your comment.

`object.send(:some_method)` is the same as `object.some_method`, but with a super dynamic twist to it. It's not always clear which method is being called because `:some_method` can actually be a variable.

So imagine:

```
method_to_call = figure_out_what_the_heck_I_want_do
my_fabulous_object.send(method_to_call)
```

You will never ever know what  the heck you're actually doing on line 2 without looking at what `figure_out_what_the_heck_I_want_do` does because the method that actually gets called on line 2 is not explicit (as in `thing.my_method`) but rather super-duper dynamic based on the result of `figure_out_what_the_heck_I_want_do`.

When you run into some clever code where somebody did this you will understand the pain.
